src/lib/logical_query_plan/lqp_utils.hpp

```diff
@@ -208,6 +208,32 @@ void visit_lqp_upwards(const std::shared_ptr<AbstractLQPNode>& lqp, Visitor visi
  */
 std::vector<std::shared_ptr<AbstractLQPNode>> lqp_find_subplan_roots(const std::shared_ptr<AbstractLQPNode>& lqp);
 
+/**
+ * Traverses @param lqp from top to bottom to find all leaf nodes.
+ * @return By default, all leaf nodes are returned.
+ *         Optionally, the template parameter can be set to return leaf nodes of type @tparam LQPNode only.
```

Comment:
Clever. Not sure if I have seen this use of template parameters before. I feel like it is somewhat surprising though. Have you considered passing in a node type as a regular argument? Admittedly, that would require the caller to cast the return values, but it might adhere better to the principle of least surprise.

I kinda like it because it is a smart hack, but I wonder if it causes more confusion that a less smart version would.

Reply:
> Have you considered passing in a node type as a regular argument?

Yes, but I did not like the verbose casting code, which led me to this "hack" ;)